cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -479,52 +512,104 @@ impl MachInstEmit for Inst {
                 debug_assert_eq!(rs1, a0());
                 debug_assert_eq!(rs2, b0());
                 let rd = allocs.next_writable(rd);
-                // Arith asserts that A * B + C = op + 2^256 * D.
-                // Now ZKVM is 256-bit and wasm max type is 64 bit, so it would never be
-                // 256 bit overflow. But in future we will need here something like:
-                // put_string(
-                //   &format!("${{_mulArith / (2 ** 64)}} => D :ARITH\n", reg_name(rd.to_reg())),
-                //    sink,
-                //);
-                // put_string(
-                //   &format!("${{_mulArith % (2 ** 64)}} => {} :ARITH\n", reg_name(rd.to_reg())),
-                //    sink,
-                //);
-                // For now we will just clear D in case it was something in it
                 put_string("0 => D\n", sink);
                 put_string("0 => C\n", sink);
                 put_string("$${var _mulArith = A * B}\n", sink);
+                put_string("${_mulArith / 18446744073709551616} => D\n", sink);
                 put_string(
-                    &format!("${{_mulArith}} => {} :ARITH\n", reg_name(rd.to_reg())),
+                    &format!(
+                        "${{_mulArith % 18446744073709551616}} => {} :ARITH\n",
```

Comment:
Looking at ZK EVM ROM, I see that we don't use modulo there https://github.com/0xPolygonHermez/zkevm-rom/blob/c17c1664a654e27de67bcbcbe6b6ae44c5b93274/main/utils.zkasm#L594

I suspect this is because it happens automatically when trying to put the value in the register. Might be worth asking Polygon team about this.

In any case, let's use bitwise & here instead of a large constant to make this code easier to read.